packages/medusa/src/services/draft-order.ts

```diff
@@ -0,0 +1,430 @@
+          this.draftOrderRepository_
+        )
+
+        const { q, ...restSelector } = selector
+        const query = buildQuery(
+          restSelector,
+          config
+        ) as FindManyOptions<DraftOrder> & ExtendedFindConfig<DraftOrder>
+
+        if (q) {
+          const where = query.where
+          delete where?.display_id
+
+          query.join = {
+            alias: "draft_order",
+            innerJoin: {
+              cart: "draft_order.cart",
+            },
+          }
+
+          query.where = (qb): void => {
+            qb.where(where)
+
+            qb.andWhere(
+              new Brackets((qb) => {
+                qb.where(`cart.email ILIKE :q`, {
+                  q: `%${q}%`,
+                }).orWhere(`draft_order.display_id::TEXT ILIKE :displayId`, {
+                  displayId: `${q}`,
+                })
+              })
+            )
+          }
+        }
+
+        return await draftOrderRepository.findAndCount(query)
+      }
+    )
+  }
+
+  /**
+   * Lists draft orders
+   * @param selector - query object for find
+   * @param config - configurable attributes for find
+   * @return list of draft orders
+   */
+  async list(
+    selector,
+    config: FindConfig<DraftOrder> = {
+      skip: 0,
+      take: 50,
+      order: { created_at: "DESC" },
+    }
+  ): Promise<DraftOrder[]> {
+    return await this.atomicPhase_(
+      async (transactionManager: EntityManager) => {
+        const draftOrderRepo = transactionManager.getCustomRepository(
+          this.draftOrderRepository_
+        )
+
+        const query = buildQuery(selector, config)
+
+        return await draftOrderRepo.find(query)
+      }
+    )
+  }
+
+  /**
+   * Creates a draft order.
+   * @param data - data to create draft order from
+   * @return the created draft order
+   */
+  async create(data: AdminPostDraftOrdersReq): Promise<DraftOrder> {
+    return await this.atomicPhase_(
+      async (transactionManager: EntityManager) => {
+        const draftOrderRepo = transactionManager.getCustomRepository(
+          this.draftOrderRepository_
+        )
+
+        if (!data.region_id) {
+          throw new MedusaError(
+            MedusaError.Types.INVALID_DATA,
+            `region_id is required to create a draft order`
+          )
+        }
+
+        if (!data.items || !data.items.length) {
+          throw new MedusaError(
+            MedusaError.Types.INVALID_DATA,
+            `Items are required to create a draft order`
+          )
+        }
+
+        const { shipping_methods, no_notification_order, items, ...rawCart } =
+          data
+
+        if (rawCart.discounts) {
+          const discounts = [...rawCart.discounts]
+          rawCart.discounts.length = 0
+
+          for (const { code } of discounts) {
+            await this.cartService_
+              .withTransaction(transactionManager)
+              .applyDiscount(rawCart as Cart, code)
+          }
```

Comment:
question: Not entirely sure why we need to extract the discounts from `rawCart` as well as set the discounts length to 0, but I might be missing or forgetting some context. Would it not suffice to do the following?
```
          for (const { code } of rawCart.discounts) {
            await this.cartService_
              .withTransaction(transactionManager)
              .applyDiscount(rawCart as Cart, code)
          }
```

Reply:
Yes indeed it should suffice, i will try to see if it still work like that but it should, I think it was just a legacy from the js version, i remember being stuck here to understand the purpose since we were extracting the discounts but still using it from the raw cart

Reply:
I ve just done the changes and see the tests first, but what was done is that each loop was resetting the rest.discounts which means the apply discount never found a matching discount on the cart.

When I look at the cart service, the apply discount was never founding the code that was given, and therefore was adding the discount on the cart reference, but then the next loop was setting the discount to [] again which. from what I can understand, will end up by only applying the last discount code. 

Does that ring a bell in the process?